ssm_jax/hmm/core.py

```diff
@@ -0,0 +1,226 @@
+                                                log_likelihoods)
+
+    # Run the sampler backward in time
+    def _step(carry, args):
+        # Unpack the inputs
+        next_state = carry
+        t, rng, filtered_probs = args
+
+        # Get parameters for time t
+        get = lambda x: x[t] if x.ndim == 3 else x
+        A = get(transition_matrix)
+
+        # Fold in the next state and renormalize
+        smoothed_probs = filtered_probs * (A @ next_state)
+        smoothed_probs /= smoothed_probs.sum()
+
+        # Sample current state
+        state = jr.choice(rng, p=smoothed_probs)
+
+        return state, state
+
+    # Run the HMM smoother
+    num_timesteps = len(log_likelihoods)
+    rngs = jr.split(rng, num_timesteps)
+    last_state = jr.choice(rngs[-1], filtered_probs[-1])
+    args = (jnp.arange(num_timesteps - 1, -1, -1),
+            rngs[:-1],
+            filtered_probs[:-1][::-1])
+    _, rev_states = lax.scan(_step, last_state, args)
+
+    # Reverse the arrays and return
+    states = jnp.row_stack([rev_states[::-1], last_state])
+    return log_normalizer, states
+
+
+def hmm_backward_filter(transition_matrix,
+                        log_likelihoods):
+    """_summary_
+
+    Args:
+        hmm_params (_type_): _description_
+        log_likelihoods (_type_): _description_
+
+    Returns:
+        _type_: _description_
+    """
+    def _step(carry, t):
+        log_normalizer, backward_pred_probs = carry
+
+        # Get parameters for time t
+        get = lambda x: x[t] if x.ndim == 3 else x
+        A = get(transition_matrix)
+        ll = log_likelihoods[t]
+
+        ll_max = ll.max()
+        backward_pred_probs = A @ (backward_pred_probs * jnp.exp(ll - ll_max))
+
+        # Renormalize
+        norm = backward_pred_probs.sum()
+        backward_pred_probs /= norm
+
+        # Update the log log_normalizer
+        log_normalizer += jnp.log(norm) + ll_max
+
+        return (log_normalizer, backward_pred_probs), backward_pred_probs
+
+    num_timesteps, num_states = log_likelihoods.shape
+    carry = (0.0, jnp.ones(num_states))
+    (log_normalizer, _), backward_pred_probs = lax.scan(
+        _step, carry, jnp.arange(num_timesteps)[::-1])
+    return log_normalizer, backward_pred_probs
+
+
+def hmm_two_filter_smoother(initial_distribution,
+                            transition_matrix,
+                            log_likelihoods):
+    # Run the filters forward and backward
+    ll, filtered_probs = hmm_filter(initial_distribution,
+                                    transition_matrix,
+                                    log_likelihoods)
+
+    _, backward_pred_probs = hmm_backward_filter(transition_matrix,
+                                                 log_likelihoods)
+
+    # Compute smoothed probabilities
+    smoothed_probs = filtered_probs * backward_pred_probs
+    norm = smoothed_probs.sum(axis=1)
+    smoothed_probs /= norm
+
+    return smoothed_probs, ll
```

Comment:
also compute the cross probabilities $P(z_t=i, z_{t+1}=j)$ and reorder the output to be consistent with `hmm_smoother` below.